addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	// if cluster is marked for deletion, then no reconciliation is needed
+	if !cluster.GetDeletionTimestamp().IsZero() {
+		return ctrl.Result{}, nil
+	}
+
+	clusterLabels := cluster.GetLabels()
+
+	// make sure the TKR object exists
+	tkrName := clusterLabels[constants.TKRLabelClassyClusters]
+	tkr, err := util.GetTKRByName(ctx, r.Client, tkrName)
+	if err != nil {
+		log.Error(err, "unable to fetch TKR object", "name", tkrName)
+		return ctrl.Result{}, err
+	}
+	// if tkr is not found, should not requeue for the reconciliation
+	if tkr == nil {
+		log.Info("TKR object not found", "name", tkrName)
+		return ctrl.Result{}, nil
+	}
+
+	// determine cluster role
+	_, isManagementCluster := clusterLabels[tkrconstants.ManagementClusterRoleLabel]
+
+	if isManagementCluster {
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		clusterClient = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.Tracker.GetClient(ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			log.Error(err, "error getting remote cluster's client")
+			return ctrl.Result{}, err
+		}
+		clusterClient = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created
+		if err := r.watchPackageInstalls(ctx, cluster, log); err != nil {
+			log.Error(err, "error watching PackageInstalls on target cluster")
+			return ctrl.Result{}, err
+		}
+		log.Info("finished setting up remote watch for the cluster")
+	}
+
+	if err := r.reconcile(ctx, clusterClient, cluster, clusterRole, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile iterates over all core/additional packages and calls reconcileClusterBootstrapStatus() for each.
+// it eventually updates ClusterBootstrapStatus with the condition entries (one for each package) in a single update operation
+// Note: use of the client c (remote or not) depends on whether the corresponding cluster is workload or management which is
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(ctx context.Context, clusterClient client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) error {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
```

Comment:
nit: I wonder if this deepcopy is needed

Reply:
@sbueringer As I'm modifying the ```clusterBootstrap``` status a bit later; was concerned if I get the following on certain circumstances  if I don't. Please correct me if wrong?
```
the object has been modified; please apply your changes to the latest version and try again
```

Reply:
You shouldn't get this error.

As far as I'm aware this line shouldn't have an effect as before you already have an instance of clusterBootstrap which is not shared anywhere else and the DeepCopy just copies it and overwrites the clusterBootstrap var.

The "object has been modified error" should occur in cases like this:
* Your controller gets the resource (which includes a resourceVersion)
* Someone else e.g. another controller modifies the resource and thus the resourceVersion in the apiserver/etcd is changed
* Your controller tries to write the change with the resourceVersion initially retrieved via Get
* apiserver detects that you try to write a resource with an old resourceVersion and returns this error

This is just how the optimistic locking in Kubernetes works and is expected. The way to handle this is that it should be just retried. In controller-runtime this usually means just returning an error and then controller-runtime automatically retries the entire reconcile and it usually works the next time (or until there are no concurrent writes on that resource).

Reply:
Many thanks @sbueringer for the very complete description here. I learned a looot! In fact, mine falls in the second category. Clusterbootstrap controller is working at the same time on ```clusterbootstrap``` object. That's why I feel like I need it still.

Reply:
I just think the DeepCopy won't do anything as it doesn't change the resourceVersion :)

Reply:
Ah I see. Ok, I'll drop DeepCopy then :)